packages/ra-core/src/auth/useLogoutIfAccessDenied.ts

```diff
@@ -76,11 +81,33 @@ const useLogoutIfAccessDenied = (): LogoutIfAccessDenied => {
                             : error && error.redirectTo
                             ? error.redirectTo
                             : undefined;
-                    logout({}, redirectTo);
+
+                    if (logoutUser) {
+                        logout({}, redirectTo);
+                    } else {
+                        const redirectToParts = redirectTo.split('?');
+                        const newLocation: LocationDescriptorObject = {
+                            pathname: redirectToParts[0],
+                        };
+
+                        if (history.location && history.location.pathname) {
+                            newLocation.state = {
+                                nextPathname: history.location.pathname,
+                                nextSearch: history.location.search,
+                            };
+                        }
+
+                        if (redirectToParts[1]) {
+                            newLocation.search = redirectToParts[1];
+                        }
+                        history.push(redirectTo);
```

Comment:
We should probably extract this from here and from `useLogout`. @fzaninotto ?

Reply:
I thought the same too, but wasn't sure where the right place for it it is. What do you both think?

Reply:
Shouldn't this be `history.push(newLocation)`? Otherwise the code above is useless.

Also, setting `nextPathName` and `nextSearch` is currently only useful for the Login page. I don't see any point in having them here. And since we don't need that, we don't need to manipulate the `redirectTo`.

Reply:
It's been a few weeks since I've gone over this code, and I'll sanity check just in case, but I believe the `history.push(redirectTo)` is the correct behaviour. In essence, we want to perform the same behaviour but as the `logout({}, redirectTo)`but to circumvent the logout functionality.

Reply:
@fzaninotto , does the video in the PR's description clarify things?

Reply:
ignore me, you're completely right!